File: packages/schemas/ts/sync.ts
/**
 * Sync envelope — exact TS mirror of `app/schemas/sync.py` (ARCHITECTURE.md §4.4)
 * and the merge contract in SYNC_STRATEGY.md §4-§6.
 */
import { EntityType, SyncOpType } from "./enums";

export interface SyncOp {
  /** `{hlc}:{entity}:{entity_id}` — idempotency key */
  op_id: string;
  entity: EntityType;
  entity_id: string;
  op: SyncOpType;
  /** CREATE: full doc · UPDATE: changed fields only · DELETE: null */
  patch: Record<string, unknown> | null;
  hlc: string;
  device_id: string;
}

export interface PushRequest {
  device_id: string;
  /** max 500 */
  ops: SyncOp[];
  last_server_seq: number;
}

export interface PushResponse {
  applied: string[];
  /** stale/duplicate (idempotent success) */
  skipped: string[];
  server_seq: number;
}

export interface ServerOp extends SyncOp {
  server_seq: number;
}

export interface PullResponse {
  ops: ServerOp[];
  next_seq: number;
  has_more: boolean;
}

/**
 * Derived fields — SYNC_STRATEGY.md §5 step 2 / §6. These are additive,
 * cross-device counters recomputed server-side from append-only facts, never
 * LWW-merged. Both `app/services/sync.py::apply_op` (server) and
 * `lib/sync/engine.ts::applyRemoteOp` (client) strip these keys from an
 * incoming op's `patch` before applying it — kept here as the single TS-side
 * source of truth so both merge call sites agree.
 */
export const DERIVED_FIELDS: Readonly<Record<EntityType, readonly string[]>> = {
  [EntityType.TASK]: ["actual_focus_seconds"],
  [EntityType.PROJECT]: [],
  [EntityType.SECTION]: [],
  [EntityType.VISION]: [],
  [EntityType.SEASON]: [],
  [EntityType.FOCUS_SESSION]: [],
  [EntityType.DAILY_REVIEW]: [],
};

/** Reserved device id for server-originated mutations (SYNC_STRATEGY.md §2). */
export const SERVER_DEVICE_ID = "server";

File: apps/api/app/schemas/sync.py
"""Sync envelope schemas — ARCHITECTURE.md §4.4; full protocol in SYNC_STRATEGY.md.

TS mirror: ``packages/schemas/ts/sync.ts``.
"""

from __future__ import annotations

from typing import Any
from uuid import UUID

from pydantic import Field

from app.schemas.base import FEBase
from app.schemas.enums import EntityType, SyncOpType


class SyncOp(FEBase):
    """One change-log entry, as produced by the client repository layer."""

    op_id: str  # f"{hlc}:{entity}:{entity_id}" — idempotency key
    entity: EntityType
    entity_id: UUID
    op: SyncOpType
    patch: dict[str, Any] | None  # CREATE: full doc · UPDATE: changed fields only · DELETE: None
    hlc: str
    device_id: str


class PushRequest(FEBase):
    """Body of ``POST /sync/push`` — a batch of unpushed local ops."""

    device_id: str
    ops: list[SyncOp] = Field(max_length=500)
    last_server_seq: int = 0


class PushResponse(FEBase):
    """Response of ``POST /sync/push``.

    ``skipped`` = stale/duplicate op_ids — idempotent success, never retried.
    """

    applied: list[str]
    skipped: list[str]
    server_seq: int


class ServerOp(SyncOp):
    """A :class:`SyncOp` as it appears in the server oplog, with its sequence number."""

    server_seq: int


class PullResponse(FEBase):
    """Response of ``GET /sync/pull``."""

    ops: list[ServerOp]
    next_seq: int
    has_more: bool
